L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp2.cc

```diff
@@ -419,6 +636,161 @@ fillMaps(const es_info& esInfo)
     l1tCellsMappingStream.close();
 }
 
+void 
+HGCalTriggerGeometryHexImp2::
+fillNeighborMaps(const es_info& esInfo)
+{
+    // Fill trigger neighbor map
+    std::ifstream l1tCellNeighborsMappingStream(l1tCellNeighborsMapping_.fullPath());
+    if(!l1tCellNeighborsMappingStream.is_open()) edm::LogError("HGCalTriggerGeometry") << "Cannot open L1TCellNeighborsMapping file\n";
+    for(std::array<char,512> buffer; l1tCellNeighborsMappingStream.getline(&buffer[0], 512); )
+    {
+        std::string line(&buffer[0]);
+        // Extract keys consisting of the wafer configuration
+        // and of the trigger cell id
+        // Match patterns (X,Y) 
+        // where X is a set of 7 bits
+        // and Y is a number with less than 4 digits
+        std::regex key_regex("\\(\\s*[01]{7}\\s*,\\s*\\d{1,3}\\s*\\)");
+        std::vector<std::string> key_tokens {
+            std::sregex_token_iterator(line.begin(), line.end(), key_regex), {}
+        };
+        if(key_tokens.size()!=1)
+        {
+            throw cms::Exception("BadGeometry")
+                << "Syntax error in the L1TCellNeighborsMapping:\n"
+                << "  Cannot find the trigger cell key in line:\n"
+                << "  '"<<&buffer[0]<<"'\n";
+        }
+        std::regex digits_regex("([01]{7})|(\\d{1,3})");
+        std::vector<std::string>  pair {
+            std::sregex_token_iterator(key_tokens[0].begin(), key_tokens[0].end(), digits_regex), {}
+        };
+        // get cell id and wafer configuration
+        int trigger_cell = std::stoi(pair[1]);
+        std::vector<int> wafer_types;
+        wafer_types.reserve(pair[0].size());
+        // Convert waferType coarse=0, fine=1 to coarse=-1, fine=1
+        for(const auto c : pair[0]) wafer_types.emplace_back( (std::stoi(std::string(&c))?1:-1) );
```

Comment:
`c`  is a `char`. So the copy should be cheap. But I can replace `auto` with `char` to make it clear that this is a loop on `char`?